pkg/propertyprovider/aks/trackers/trackers.go

```diff
@@ -0,0 +1,570 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package trackers feature implementations that help track specific stats about
+// Kubernetes resources, e.g., nodes and pods in the AKS property provider.
+package trackers
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// AKSClusterNodeSKULabelName is the node label added by AKS, which indicated the SKU
+	// of the node.
+	AKSClusterNodeSKULabelName = "beta.kubernetes.io/instance-type"
+)
+
+// supportedResourceNames is a list of resource names that the AKS property provider supports.
+//
+// Currently the supported resources are CPU and memory.
+var supportedResourceNames []corev1.ResourceName = []corev1.ResourceName{
+	corev1.ResourceCPU,
+	corev1.ResourceMemory,
+}
+
+// NodeSet is a set of nodes.
+type NodeSet map[string]bool
+
+// NodeTracker helps track specific stats about nodes in a Kubernetes cluster, e.g., its count.
+type NodeTracker struct {
+	// totalCapacity and totalAllocatable are the total capacity and allocatable capacity
+	// of the cluster, respectively.
+	totalCapacity    corev1.ResourceList
+	totalAllocatable corev1.ResourceList
+	// perCPUCoreHourlyCost and perGBMemoryHourlyCost are the average per CPU core and per GB memory
+	// costs in the cluster, respectively.
+	perCPUCoreHourlyCost  float64
+	perGBMemoryHourlyCost float64
+
+	// Below are a list of maps that tracks information about individual nodes in the cluster.
+	capacityByNode    map[string]corev1.ResourceList
+	allocatableByNode map[string]corev1.ResourceList
+	nodeSetBySKU      map[string]NodeSet
+	skuByNode         map[string]string
+
+	// costLastUpdated is the timestamp when the per-resource-unit costs are last calculated.
+	costLastUpdated time.Time
+	// costErr tracks any error that occurs during cost calculation.
+	costErr error
```

Comment:
do you mean the perXXX fields?

create a separate struct to group them together?

Reply:
Hi Zhiying! You mean the error part? Only cost has that, and it's not one for each cost property, to streamline the experience.

Reply:
it's about line 55 comment, "costLastUpdated is the timestamp when the per-resource-unit costs are last calculated."

I assume both costLastUpdate and costErr are all talking about the "per-resource-unit costs".

Will it be better to be grouped into a separate structure, eg

type struct PerResourceCost {
     perCPUCoreHourlyCost  float64
     perGBMemoryHourlyCost float64
     costLastUpdated time.Time
	// costErr tracks any error that occurs during cost calculation.
	costErr error
}

Reply:
Discussed offline.